cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramNotificationSubscriberService.java

```diff
@@ -0,0 +1,249 @@
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+
+/**
+ * Service that receives program status notifications and persists to the store
+ */
+public class ProgramNotificationSubscriberService extends AbstractNotificationSubscriberService {
+  private static final Logger LOG = LoggerFactory.getLogger(ProgramNotificationSubscriberService.class);
+  private static final Gson GSON = new Gson();
+  private static final Type STRING_STRING_MAP = new TypeToken<Map<String, String>>() { }.getType();
+
+  private final CConfiguration cConf;
+  private ExecutorService taskExecutorService;
+
+  @Inject
+  ProgramNotificationSubscriberService(MessagingService messagingService, CConfiguration cConf,
+                                       DatasetFramework datasetFramework, TransactionSystemClient txClient) {
+    super(messagingService, cConf, datasetFramework, txClient);
+    this.cConf = cConf;
+  }
+
+  @Override
+  protected void startUp() {
+    LOG.info("Starting ProgramNotificationSubscriberService");
+    taskExecutorService = Executors.newCachedThreadPool(new ThreadFactoryBuilder()
+                                                          .setNameFormat("program-status-subscriber-task-%d")
+                                                          .build());
+    taskExecutorService.submit(new ProgramStatusNotificationSubscriberThread(
+      cConf.get(Constants.AppFabric.PROGRAM_STATUS_EVENT_TOPIC)));
+  }
+
+  @Override
+  protected void shutDown() {
+    super.shutDown();
+    try {
+      taskExecutorService.awaitTermination(5, TimeUnit.SECONDS);
+    } catch (InterruptedException ie) {
+      Thread.currentThread().interrupt();
+    } finally {
+      if (!taskExecutorService.isTerminated()) {
+        taskExecutorService.shutdownNow();
+      }
+    }
+    LOG.info("Stopped ProgramNotificationSubscriberService.");
+  }
+
+  /**
+   * Thread that receives TMS notifications and persists the program status notification to the store
+   */
+  private class ProgramStatusNotificationSubscriberThread extends NotificationSubscriberThread {
+    private final String topic;
+
+    ProgramStatusNotificationSubscriberThread(String topic) {
+      super(topic);
+      this.topic = topic;
+    }
+
+    @Override
+    public String loadMessageId(DatasetContext context) {
+      return getAppMetadataStore(context).retrieveSubscriberState(topic);
+    }
+
+    @Override
+    public void updateMessageId(DatasetContext context, String lastFetchedMessageId) {
+      getAppMetadataStore(context).persistSubscriberState(topic, lastFetchedMessageId);
+    }
+
+    @Override
+    public void processNotification(final DatasetContext context, Notification notification)
+      throws IOException, DatasetManagementException {
+      Map<String, String> properties = notification.getProperties();
+      // Required parameters
+      String programRunIdString = properties.get(ProgramOptionConstants.PROGRAM_RUN_ID);
```

Comment:
nit: better rename to runid. we usually do not add type information in the variable name.